sky/skylet/providers/lambda_labs/node_provider.py

```diff
@@ -0,0 +1,193 @@
+import logging
+import os
+import time
+from threading import RLock
+
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import TAG_RAY_CLUSTER_NAME
+from sky.authentication import PRIVATE_SSH_KEY_PATH
+from sky.skylet.providers.lambda_labs.lambda_utils import LambdaLabsClient, Metadata
+from sky.utils import command_runner
+
+REMOTE_TAG_PATH_PREFIX = '/home/ubuntu/.lambda-metadata'
+LOCAL_TAG_PATH_PREFIX = '~/.lambda_labs/metadata'
+
+logger = logging.getLogger(__name__)
+
+
+def _on_local():
+    return os.path.exists(os.path.expanduser(PRIVATE_SSH_KEY_PATH))
+
+
+def _send_tags_file(ip, cluster_name):
+    if not _on_local():
+        return
+
+    # Local
+    path = os.path.expanduser(f'{LOCAL_TAG_PATH_PREFIX}-{cluster_name}')
+    runner = command_runner.SSHCommandRunner(ip, 'ubuntu', PRIVATE_SSH_KEY_PATH)
+    runner.rsync(path, f'{REMOTE_TAG_PATH_PREFIX}-{cluster_name}', up=True)
+
+
+def synchronized(f):
+    def wrapper(self, *args, **kwargs):
+        self.lock.acquire()
+        try:
+            return f(self, *args, **kwargs)
+        finally:
+            self.lock.release()
+
+    return wrapper
+
+
+class LambdaNodeProvider(NodeProvider):
+    """Node Provider for Lambda Labs.
+
+    Lambda credentials must be set up.
+
+    Nodes may be in one of three states: {pending, running, terminated}. Nodes
+    appear immediately once started by ``create_node``, and transition
+    immediately to terminated when ``terminate_node`` is called.
+    """
+
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+        self.lock = RLock()
+        self.lambda_client = LambdaLabsClient()
+        # Only used for tags
+        if _on_local():
```

Comment:
Yes, it's because for auto{down, stop} to work, you have to be able to run `ray up/down` in the head node. Since we are storing ray tags locally, we have also store ray tags on the head node.